cartridges/sg_changes/cartridge/controllers/COPlaceOrder.js

```diff
@@ -0,0 +1,267 @@
+            };
+        }
+        /**
+         * Sets the transaction ID for the payment instrument.
+         */
+        var handlePaymentTransaction = function () {
+            paymentInstrument.getPaymentTransaction().setTransactionID(order.getOrderNo());
+        };
+
+        for (var i = 0; i < paymentInstruments.length; i++) {
+            var paymentInstrument = paymentInstruments[i];
+
+            if (PaymentMgr.getPaymentMethod(paymentInstrument.getPaymentMethod()).getPaymentProcessor() === null) {
+
+                Transaction.wrap(handlePaymentTransaction);
+
+            } else {
+
+                var authorizationResult = PaymentProcessor.authorize(order, paymentInstrument);
+
+                if (authorizationResult.not_supported || authorizationResult.error) {
+                    return {
+                        error: true
+                    };
+                }
+            }
+        }
+    }
+
+    return {};
+}
+
+/**
+ * The entry point for order creation. This function is not exported, as this controller must only
+ * be called by another controller.
+ *
+ * @transactional
+ * @return {Object} JSON object that is empty, contains error information, or PlaceOrderError status information.
+ */
+function start() {
+    var cart = Cart.get();
+
+    if (!cart) {
+        app.getController('Cart').Show();
+        return {};
+    }
+
+    var COShipping = app.getController('COShipping');
+
+    // Clean shipments.
+    COShipping.PrepareShipments(cart);
+
+    // Make sure there is a valid shipping address, accounting for gift certificates that do not have one.
+    if (cart.getProductLineItems().size() > 0 && cart.getDefaultShipment().getShippingAddress() === null) {
+        COShipping.Start();
+        return {};
+    }
+
+    // Make sure the billing step is fulfilled, otherwise restart checkout.
+    if (!session.forms.billing.fulfilled.value) {
+        app.getController('COCustomer').Start();
+        return {};
+    }
+
+    Transaction.wrap(function () {
+        cart.calculate();
+    });
+
+    var COBilling = app.getController('COBilling');
+
+    Transaction.wrap(function () {
+        if (!COBilling.ValidatePayment(cart)) {
+            COBilling.Start();
+            return {};
+        }
+    });
+
+    // Recalculate the payments. If there is only gift certificates, make sure it covers the order total, if not
+    // back to billing page.
+    Transaction.wrap(function () {
+        if (!cart.calculatePaymentTransactionTotal()) {
+            COBilling.Start();
+            return {};
+        }
+    });
+
+    // Handle used addresses and credit cards.
+    var saveCCResult = COBilling.SaveCreditCard();
+
+    if (!saveCCResult) {
+        return {
+            error: true,
+            PlaceOrderError: new Status(Status.ERROR, 'confirm.error.technical')
+        };
+    }
+
+    var affirmController = require('int_affirm_controllers/cartridge/controllers/Affirm');
```

Comment:
Why isn't this using the asterisk reference?

Reply:
This is a controller reference specific to the SiteGenesis controllers cartridge. Astound included only the controllers code for the SiteGenesis reference files (under `/cartridges/sg_changes/`).

Reply:
When do you use the asterisk vs using the absolute path?

Reply:
- Current cartridge relative to the file location: `require('~/cartridge/scripts/cart')`
- From the beginning of the cartridge path: `require('*/cartridge/scripts/util/array')`
where cartridge path, as defined in BM, get searched from right to left. 
If the cartridge path looks like below, and and the code is in int_affirm_pipelines, it will first search base followed by int_affirm and then itself which is int_affirm_pipelines
`int_affirm_pipelines;int_affirm;base`

Reply:
Okay cool, as long as it's consistent when one type is used vs the other :)